packages/connect/src/createScClient/createScClient.ts

```diff
@@ -11,19 +20,293 @@ export interface ScClient {
+  > = new Map()
+  readonly #orphanMessages: Map<string, Array<string | Error>> = new Map()
+  readonly #requests: Map<number, ResponseCallback> = new Map()
+  readonly #eventemitter: EventEmitter = new EventEmitter()
+  #chain: Promise<Chain> | null = null
+  #isChainReady: boolean = false
+
+  public constructor(getChain: (handler: JsonRpcCallback) => Promise<Chain>) {
+    this.#getChain = getChain
+  }
+
+  public get hasSubscriptions(): boolean {
+    // you would expect this to be `this.#subscriptions.size > 0`, right?
+    // well, nope... This must always return `true` b/c otherwise Polkadot
+    // thinks that this Provider doesn't support subscriptions.
+    return true
+  }
+
+  public get isConnected(): boolean {
+    return !!this.#chain && this.#isChainReady
+  }
+
+  public clone(): ProviderInterface {
+    throw new Error("clone() is not supported.")
+  }
+
+  async connect(): Promise<void> {
+    if (this.isConnected) throw new Error("Already connected!")
+
+    // it could happen that after emitting `disconnected` due to the fact taht
+    // smoldot is syncing, the consumer tries to reconnect after a certain amount
+    // of time... In which case we want to make sure that we don't create a new
+    // chain.
+    if (this.#chain) {
+      await this.#chain
+      return
+    }
+
+    const hc = healthChecker()
+    const onResponse = (res: string): void => {
+      if (!hc.responsePassThrough(res)) {
+        return
+      }
+
+      const response = JSON.parse(res) as JsonRpcResponse
+
+      // the healthChecker alters the original id. So, we must restore it
+      // because otherwise the RpcCoder won't be able to decode the resonse.
+      if (
+        typeof response.id === "string" &&
+        (response.id as unknown as string).startsWith("extern:")
+      ) {
+        response.id = Number((response.id as unknown as string).slice(7))
+      }
+
+      let decodedResponse: string | Error
+      try {
+        decodedResponse = this.#coder.decodeResponse(response) as string
+      } catch (e) {
+        decodedResponse = e as Error
+      }
+
+      // It's not a subscription message, but rather a standar RPC response
+      if (response.method === undefined) {
+        return this.#requests.get(response.id)?.(decodedResponse)
+      }
+
+      // It has a `method` property, that means that it's a subscription message
+      const subscriptionId = `${response.method}::${response.params.subscription}`
+
+      const callback = this.#subscriptions.get(subscriptionId)?.[0]
+      if (callback) return callback(decodedResponse)
+
+      // It's possible to receive subscriptions messages before having received
+      // the id of the subscription. In that case We should keep these
+      // messages around until we receive the subscription-id message.
+      if (!this.#orphanMessages.has(subscriptionId))
+        this.#orphanMessages.set(subscriptionId, [])
+      this.#orphanMessages.get(subscriptionId)!.push(decodedResponse)
```

Comment:
Smoldot will always send back the id before the first messages.
You're reaching this because you're using callbacks in the `subscribe` methods, meaning that the response containing the "id" is processed later despite having arrived before.

Reply:
> You're reaching this because you're using callbacks in the subscribe methods, meaning that the response containing the "id" is processed later despite having arrived before.

Nah... The reason why I'm doing this is because this is what happens over the wire when using the WsProvider... I thought that smoldot was going to do the same thing.

Reply:
> You're reaching this because you're using callbacks in the subscribe methods, meaning that the response containing the "id" is processed later despite having arrived before.

Nah... The reason why I'm doing this is because this is what happens over the wire when using the WsProvider... I thought that smoldot was going to do the same thing.

Reply:
I guess that we should keep this code for now, at least until [this bug](https://github.com/paritytech/substrate/issues/10910) is fixed.

Reply:
Why? The code in this PR will never interact with Substrate

Reply:
Oh! My bad... So, what you are saying is that the smoldot  chain doesn't have this issue? That's awesome! Well, then in that case I will update the PR...

Reply:
[Done](https://github.com/paritytech/substrate-connect/pull/787/commits/f6bd57eaddcf8cb95834abd0b83229acc5e74145)!

FTR those changes and their tests should be a pretty convincing prove that it's never been the case that the code of this PR was:

> reaching this because you're using callbacks in the subscribe methods, meaning that the response containing the "id" is processed later despite having arrived before